pkg/storageincentives/incentives_test.go

```diff
@@ -0,0 +1,230 @@
+// Copyright 2022 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package storageincentives_test
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/log"
+	"github.com/ethersphere/bee/pkg/postage"
+	mockbatchstore "github.com/ethersphere/bee/pkg/postage/batchstore/mock"
+	"github.com/ethersphere/bee/pkg/storageincentives"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/swarm/test"
+	"go.uber.org/atomic"
+)
+
+func Test(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name           string
+		blocksPerRound float64
+		blocksPerPhase float64
+		incrementBy    float64
+		limit          float64
+		expectedCalls  int
+	}{{
+		name:           "3 blocks per phase, same block number returns twice",
+		blocksPerRound: 9,
+		blocksPerPhase: 3,
+		incrementBy:    1,
+		expectedCalls:  10,
+		limit:          108, // computed with blocksPerRound * (exptectedCalls + 2)
+	}, {
+		name:           "3 blocks per phase, block number returns every block",
+		blocksPerRound: 9,
+		blocksPerPhase: 3,
+		incrementBy:    1,
+		expectedCalls:  10,
+		limit:          108,
+	}, {
+		name:           "3 blocks per phase, block number returns every other block",
+		blocksPerRound: 9,
+		blocksPerPhase: 3,
+		incrementBy:    2,
+		expectedCalls:  10,
+		limit:          108,
+	}, {
+		name:           "no expected calls - block number returns late after each phase",
+		blocksPerRound: 9,
+		blocksPerPhase: 3,
+		incrementBy:    6,
+		expectedCalls:  0,
+		limit:          108,
+	}, {
+		name:           "4 blocks per phase, block number returns every other block",
+		blocksPerRound: 12,
+		blocksPerPhase: 4,
+		incrementBy:    2,
+		expectedCalls:  10,
+		limit:          144,
+	},
+	}
+
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			wait := make(chan struct{})
+			addr := test.RandomAddress()
+
+			backend := &mockchainBackend{
+				limit: tc.limit,
+				limitCallback: func() {
+					select {
+					case wait <- struct{}{}:
+					default:
+					}
+				},
+				incrementBy: tc.incrementBy,
+				block:       tc.blocksPerRound}
+			contract := &mockContract{t: t, baseAddr: addr}
+
+			service := createService(addr, backend, contract, uint64(tc.blocksPerRound), uint64(tc.blocksPerPhase))
+
+			<-wait
+
+			time.Sleep(time.Millisecond * 500)
+
+			if int(contract.commitCalls.Load()) != tc.expectedCalls {
+				t.Fatalf("got %d, want %d", contract.commitCalls.Load(), tc.expectedCalls)
+			}
+
+			if int(contract.revealCalls.Load()) != tc.expectedCalls {
```

Comment:
The test does not check the sample/sampleEnd events?

Also, it would be good to test both the cases of commit. I think right now only one path is tested where you get to commit and you already have sampleEnd.